File: backend/app/routers/docs.py
"""API 文档发布路由 — 发布、分享、公开访问、文档编辑、版本管理"""

from typing import Optional
from fastapi import APIRouter, Depends, Request
from pydantic import BaseModel, Field
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.limiter import limiter
from app.middleware.auth import get_current_user, get_optional_user
from app.models.user import User
from app.models.project import Project
from app.services.doc_service import DocService
from app.services.permission_service import check_read_access, check_write_access
from app.utils.response import success

router = APIRouter(prefix="/projects/{project_id}/docs", tags=["API Documentation"])


class DocPublishRequest(BaseModel):
    name: str = Field(default="", description="文档名称")
    description: str = Field(default="", description="文档描述")
    password: str = Field(default="", description="访问密码（可选）")
    expires_in_days: int = Field(default=0, description="过期天数，0 表示永不过期")
    include_categories: list[int] = Field(default=[], description="包含的接口目录 ID 列表，空=全部")
    env_id: Optional[int] = Field(default=None, description="关联环境 ID（可选）")


@router.post("/publish", summary="发布文档", description="将项目 API 发布为在线文档，生成分享链接")
async def publish_docs(
    project_id: int,
    req: DocPublishRequest,
    current_user: User = Depends(get_current_user),
    _project: Project = Depends(check_write_access),
    db: AsyncSession = Depends(get_db),
):
    """发布 API 文档"""
    s = DocService(db)
    doc = await s.create(
        project_id=project_id,
        user_id=current_user.id,
        name=req.name or f"{_project.name} API 文档",
        description=req.description,
        password=req.password,
        expires_in_days=req.expires_in_days,
        include_categories=req.include_categories,
        env_id=req.env_id,
    )
    return success({
        **s.to_dict(doc),
        "share_url": f"/shared-docs/{doc.share_token}",
    })


@router.get("/list", summary="文档列表", description="获取项目的已发布文档列表")
async def list_docs(
    project_id: int,
    current_user: User | None = Depends(get_optional_user),
    _project: Project = Depends(check_read_access),
    db: AsyncSession = Depends(get_db),
):
    s = DocService(db)
    return success(await s.list_by_project(project_id))


@router.post("/{doc_id}/revoke", summary="撤销文档", description="撤销已发布的 API 文档分享链接")
async def revoke_doc(
    project_id: int,
    doc_id: int,
    current_user: User = Depends(get_current_user),
    _project: Project = Depends(check_write_access),
    db: AsyncSession = Depends(get_db),
):
    s = DocService(db)
    await s.revoke(doc_id, project_id)
    return success(message="文档已撤销")


@router.delete("/{doc_id}", summary="删除文档", description="删除文档发布记录")
async def delete_doc(
    project_id: int,
    doc_id: int,
    current_user: User = Depends(get_current_user),
    _project: Project = Depends(check_write_access),
    db: AsyncSession = Depends(get_db),
):
    s = DocService(db)
    await s.delete(doc_id, project_id)
    return success(message="文档已删除")


# ── API 文档编辑端点 ──────────────────────────────────────────────

class ApiDocSaveRequest(BaseModel):
    description: str = Field(default="", description="Markdown 格式的接口描述")
    param_docs: list[dict] | None = Field(default=None, description="参数文档列表 [{name, type, required, description}]")
    request_examples: list[dict] | None = Field(default=None, description="请求示例列表 [{name, body}]")
    response_examples: list[dict] | None = Field(default=None, description="响应示例列表 [{name, status_code, body}]")
    is_draft: bool = Field(default=True, description="是否为草稿，false 时创建版本快照")


@router.get("/api/{api_id}", summary="获取接口文档", description="获取单个 API 的文档数据")
async def get_api_doc(
    project_id: int,
    api_id: int,
    current_user: User | None = Depends(get_optional_user),
    _project: Project = Depends(check_read_access),
    db: AsyncSession = Depends(get_db),
):
    s = DocService(db)
    return success(await s.get_api_doc(project_id, api_id))


@router.put("/api/{api_id}", summary="保存接口文档", description="保存单个 API 的文档数据，支持草稿和发布")
async def save_api_doc(
    project_id: int,
    api_id: int,
    req: ApiDocSaveRequest,
    current_user: User = Depends(get_current_user),
    _project: Project = Depends(check_write_access),
    db: AsyncSession = Depends(get_db),
):
    s = DocService(db)
    result = await s.save_api_doc(
        project_id=project_id,
        api_id=api_id,
        user_id=current_user.id,
        description=req.description,
        param_docs=req.param_docs,
        request_examples=req.request_examples,
        response_examples=req.response_examples,
        is_draft=req.is_draft,
    )
    return success(result, message="草稿已保存" if req.is_draft else "文档已发布")


# ── 文档版本管理端点 ──────────────────────────────────────────────

@router.get("/api/{api_id}/versions", summary="文档版本历史", description="获取 API 文档的版本历史列表")
async def list_doc_versions(
    project_id: int,
    api_id: int,
    current_user: User | None = Depends(get_optional_user),
    _project: Project = Depends(check_read_access),
    db: AsyncSession = Depends(get_db),
):
    s = DocService(db)
    return success(await s.list_versions(api_id))


@router.post("/api/{api_id}/versions/{version_id}/rollback", summary="回滚文档版本", description="回滚到指定的文档版本")
async def rollback_doc_version(
    project_id: int,
    api_id: int,
    version_id: int,
    current_user: User = Depends(get_current_user),
    _project: Project = Depends(check_write_access),
    db: AsyncSession = Depends(get_db),
):
    s = DocService(db)
    result = await s.rollback_version(api_id, version_id, current_user.id)
    return success(result, message="已回滚到指定版本")


# 公开共享文档路由（无需认证，放在公共前缀下）
shared_router = APIRouter(prefix="/shared", tags=["Shared Docs"])


class DocAccessRequest(BaseModel):
    password: str = Field(default="", description="文档访问密码")


@shared_router.get("/docs/{token}")
@limiter.limit("60/minute")
async def get_shared_doc(
    token: str,
    request: Request,
    db: AsyncSession = Depends(get_db),
):
    """通过分享令牌获取 API 文档数据（公开，无需登录，仅限无密码文档）"""
    s = DocService(db)
    result = await s.get_by_token(token, "")
    return success(result)


@shared_router.post("/docs/{token}")
@limiter.limit("60/minute")
async def get_shared_doc_with_password(
    token: str,
    request: Request,
    body: DocAccessRequest,
    db: AsyncSession = Depends(get_db),
):
    """通过分享令牌和密码获取 API 文档数据（公开，无需登录，用于密码保护文档）"""
    s = DocService(db)
    result = await s.get_by_token(token, body.password)
    return success(result)

File: frontend/src/api/docs.ts
import request from './request'

export interface DocPublishRequest {
  name?: string
  description?: string
  password?: string
  expires_in_days?: number
  include_categories?: number[]
  env_id?: number | null
}

export interface DocItem {
  id: number
  name: string
  description: string
  project_id: number
  share_token: string
  password_hash: string
  expires_at: string | null
  is_revoked: boolean
  created_by: number
  created_at: string
  updated_at: string
}

export interface DocPublishResponse extends DocItem {
  share_url: string
}

export interface SharedDocData {
  project_name: string
  description: string
  categories: unknown[]
  apis: unknown[]
  environment?: unknown
}

export interface ParamDoc {
  name: string
  type: string
  required: boolean
  description: string
}

export interface RequestExample {
  name: string
  body: string
}

export interface ResponseExample {
  name: string
  status_code: number | string
  body: string
}

export interface ApiDocData {
  api_id: number
  name: string
  method: string
  path: string
  description: string
  description_md: string
  param_docs: ParamDoc[]
  request_examples: RequestExample[]
  response_examples: ResponseExample[]
  is_draft: boolean
}

export interface ApiDocSaveRequest {
  description?: string
  param_docs?: ParamDoc[] | null
  request_examples?: RequestExample[] | null
  response_examples?: ResponseExample[] | null
  is_draft?: boolean
}

export interface DocVersionItem {
  id: number
  api_id: number
  change_summary: string
  changed_by: number | null
  created_at: string
}

// ── 发布 API 文档 ──
export function publishDocs(projectId: number, data: DocPublishRequest) {
  return request.post<DocPublishResponse>(`/projects/${projectId}/docs/publish`, data)
}

// ── 文档列表 ──
export function listDocs(projectId: number) {
  return request.get<DocItem[]>(`/projects/${projectId}/docs/list`)
}

// ── 撤销文档 ──
export function revokeDoc(projectId: number, docId: number) {
  return request.post(`/projects/${projectId}/docs/${docId}/revoke`)
}

// ── 删除文档 ──
export function deleteDoc(projectId: number, docId: number) {
  return request.delete(`/projects/${projectId}/docs/${docId}`)
}

// ── 通过分享令牌获取文档（公开）──
export function getSharedDoc(token: string, password?: string) {
  return request.get<SharedDocData>(`/shared/docs/${token}`, {
    params: { password },
  })
}

// ── 获取接口文档 ──
export function getApiDoc(projectId: number, apiId: number) {
  return request.get<ApiDocData>(`/projects/${projectId}/docs/api/${apiId}`)
}

// ── 保存接口文档 ──
export function saveApiDoc(projectId: number, apiId: number, data: ApiDocSaveRequest) {
  return request.put<ApiDocData>(`/projects/${projectId}/docs/api/${apiId}`, data)
}

// ── 获取文档版本历史 ──
export function listDocVersions(projectId: number, apiId: number) {
  return request.get<DocVersionItem[]>(`/projects/${projectId}/docs/api/${apiId}/versions`)
}

// ── 回滚文档版本 ──
export function rollbackDocVersion(projectId: number, apiId: number, versionId: number) {
  return request.post<ApiDocData>(`/projects/${projectId}/docs/api/${apiId}/versions/${versionId}/rollback`)
}